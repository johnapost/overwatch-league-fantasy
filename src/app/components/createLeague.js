// @flow

import React, { Component } from "react";
import { connect } from "react-redux";
import { compose } from "redux";
import { Button, Input, Form } from "antd";
import uuid from "uuid/v4";
import rosterSlots from "../shared/roster";
import withFirestore from "../shared/withFirestore";

import type { StoreState } from "../shared/makeStore";
import type { UserState } from "../redux/user";

type Props = {
  firebase: Object,
  firestore: Object,
  form: Object,
  user: UserState
};

type State = {
  checkingLeague: boolean
};

export class CreateLeagueComponent extends Component<Props, State> {
  state = {
    checkingLeague: false
  };

  checkLeague = () => {
    const {
      form: { getFieldsValue, setFields },
      firestore
    } = this.props;
    const { leagueName } = getFieldsValue();
    if (!leagueName) return;

    this.setState({ checkingLeague: true });
    firestore
      .get({
        collection: "leagues",
        where: [["name", "==", leagueName]]
      })
      .then(({ empty }) => {
        if (!empty) {
          setFields({
            leagueName: {
              value: leagueName,
              errors: [new Error(`${leagueName} is already taken!`)]
            }
          });
          this.setState({ checkingLeague: false });
          return;
        }
        setFields({
          leagueName: {
            value: leagueName,
            errors: null
          }
        });
        this.createLeague(leagueName);
      });
  };

  createLeague = async (name: string) => {
    const {
      firebase: { firestore: firestoreDep },
      user: { uid }
    } = this.props;
    const db = firestoreDep();
    const batch = db.batch();
    const leagueId = uuid();
    batch.set(db.collection("leagues").doc(leagueId), {
      drafter: null,
      leagueUsers: [uid],
      name,
      ownerUser: uid,
      rosterSlots
    });
    batch.update(db.collection("users").doc(uid), {
      userLeagues: [leagueId]
    });
    return batch.commit();
  };

  render() {
    const { checkingLeague } = this.state;
    const {
      form: { getFieldDecorator }
    } = this.props;

    return (
      <Form layout="inline">
        <Form.Item>
          {getFieldDecorator("leagueName")(
            <Input
              placeholder="League name"
              onPressEnter={checkingLeague ? () => {} : this.checkLeague}
            />
          )}
        </Form.Item>
        <Form.Item>
          <Button disabled={checkingLeague} onClick={this.checkLeague}>
            Create a league
          </Button>
        </Form.Item>
      </Form>
    );
  }
}

export const mapStateToProps = ({ user }: StoreState) => ({ user });

export default compose(
  connect(mapStateToProps),
  withFirestore(),
  Form.create()
)(CreateLeagueComponent);
