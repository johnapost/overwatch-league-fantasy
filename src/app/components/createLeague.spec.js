import React from "react";
import { shallow } from "enzyme";
import { Button, Input } from "antd";
import { CreateLeagueComponent } from "./createLeague";

describe("CreateLeague", () => {
  describe("component", () => {
    const props = {
      form: {
        getFieldDecorator: jest.fn(() => jest.fn(component => component)),
        getFieldsValue: jest.fn(),
        setFields: jest.fn()
      },
      firestore: {
        get: jest.fn(),
        set: jest.fn(),
        then: jest.fn()
      },
      user: {
        uid: "abc123"
      }
    };

    it("should render fine", () => {
      const wrapper = shallow(<CreateLeagueComponent {...props} />);
      expect(wrapper.html()).toMatchSnapshot();
    });

    describe("when checking league availability", () => {
      it("should disable submission", () => {
        const wrapper = shallow(<CreateLeagueComponent {...props} />);
        wrapper.setState({ checkingLeague: true });
        expect(wrapper.find(Button).prop("disabled")).toBe(true);
        expect(wrapper.find(Input).prop("onPressEnter")).not.toBe(
          wrapper.instance().checkLeague
        );
      });
    });
  });
});
