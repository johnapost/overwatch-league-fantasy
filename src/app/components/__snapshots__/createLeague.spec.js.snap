// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateLeague component should render fine 1`] = `"<form class=\\"ant-form ant-form-inline\\"><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><input type=\\"text\\" placeholder=\\"League name\\" value=\\"\\" class=\\"ant-input\\"/></span></div></div></div><button type=\\"button\\" class=\\"ant-btn\\"><span>Create a league</span></button></form>"`;
